LanguageFeatures/Patterns/execution_pattern_for_in_element_A01_t01.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// A collection element of the form:
+///
+/// for (<keyword> <pattern> in <expression>) <element>
+///
+/// Where <keyword> is var or final is treated like so:
+/// 1. Let I be the static type of <expression>, inferred using context type
+///   schema Iterable<P> where P is the context type schema of <pattern>.
+/// 2. Calculate the element type of I:
+///     i. If I implements Iterable<T> for some T then E is T.
+///     ii. Else if I is dynamic then E is dynamic.
+///     iii. Else it is a compile-time error.
+/// 3. Type check <pattern> with matched value type E.
+/// 4. If there are no compile-time errors, then execution proceeds as the
+///   following code, where id1 and id2 are fresh identifiers and append() is
+///   an operation to add an element to the surrounding collection being built:
+/// ```dart
+/// I id1 = <expression>;
+/// Iterator<E> id2 = id1.iterator;
+/// while (id2.moveNext()) {
+///   <keyword> <pattern> = id2.current;
+///   append(<element>);
+/// }
+/// ```
+/// @description Checks that if `I` implements `Iterable<T>` for some `T` then
+/// `E` is `T`. Test `T` inferred from `<pattern>` to `<expression>` and from
+/// `<expression>` to `<pattern>`
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns,records
+
+import "../../Utils/expect.dart";
+import "../../Utils/static_type_helper.dart";
+import "patterns_lib.dart";
+
+main() {
+  var l1 = [
+    0,
+    for (var (num v1) in [1, 2, 3]) v1,
+    4
+  ];
+  Expect.listEquals([0, 1, 2, 3, 4], l1);
+  l1.expectStaticType<Exactly<List<num>>>();
```

Comment:
Yes, that would be a bug: When it's `List<dynamic>` I suspect that the context type is `_`, but it should have been `Iterable<num>`, because the type schema of the pattern is `num`.